app/assets/javascripts/discourse/app/components/d-multi-select.gjs

```diff
@@ -66,87 +101,104 @@ export default class DMultiSelect extends Component {
+        this.toggle(this.preselectedItem, event);
       }
     }
 
     if (event.key === "ArrowDown") {
       event.preventDefault();
 
-      if (!this.results || this.results.length === 0) {
+      if (!this.data.value?.length) {
         return;
       }
 
       if (this.preselectedItem === null) {
-        this.preselectedItem = this.results[0];
+        this.preselectedItem = this.data.value[0];
       } else {
-        const currentIndex = this.results.findIndex((item) =>
+        const currentIndex = this.data.value.findIndex((item) =>
           this.compare(item, this.preselectedItem)
         );
 
-        if (currentIndex < this.results.length - 1) {
-          this.preselectedItem = this.results[currentIndex + 1];
+        if (currentIndex < this.data.value.length - 1) {
+          this.preselectedItem = this.data.value[currentIndex + 1];
         }
       }
     }
 
     if (event.key === "ArrowUp") {
       event.preventDefault();
 
-      if (!this.results || this.results.length === 0) {
+      if (!this.data.value?.length) {
         return;
       }
 
       if (this.preselectedItem === null) {
-        this.preselectedItem = this.results[0];
+        this.preselectedItem = this.data.value[0];
       } else {
-        const currentIndex = this.results.findIndex((item) =>
+        const currentIndex = this.data.value.findIndex((item) =>
           this.compare(item, this.preselectedItem)
         );
 
         if (currentIndex > 0) {
-          this.preselectedItem = this.results[currentIndex - 1];
+          this.preselectedItem = this.data.value[currentIndex - 1];
         }
       }
     }
   }
 
   @action
   remove(selectedItem, event) {
-    event.stopPropagation();
+    event?.stopPropagation();
 
-    this.selection = new TrackedArray(
-      this.selection.filter((item) => !this.compare(item, selectedItem))
+    this.args.onChange?.(
+      this.args.selection?.filter((item) => !this.compare(item, selectedItem))
     );
   }
 
   @action
   isSelected(result) {
-    return this.selection.filter((item) => this.compare(item, result)).length;
+    return this.args.selection?.filter((item) => this.compare(item, result))
+      .length;
   }
 
   @action
   toggle(result, event) {
     event?.stopPropagation();
 
     if (this.isSelected(result)) {
-      this.selection = new TrackedArray(
-        this.selection.filter((item) => !this.compare(item, result))
-      );
+      this.remove(result, event);
     } else {
-      this.selection.push(result);
+      this.args.onChange?.(makeArray(this.args.selection).concat(result));
     }
-
-    this.args.onChange?.(this.selection);
   }
 
   @action
   compare(a, b) {
-    return this.args.compareFn?.(a, b) ?? a.id === b.id;
+    if (this.args.compareFn) {
+      return this.args.compareFn(a, b);
+    } else {
+      return a[this.compareKey] === b[this.compareKey];
+    }
+  }
+
+  #isPromise(value) {
+    return value instanceof Promise || value instanceof RsvpPromise;
+  }
```

Comment:
A common pattern for this is to check whether something [is a "thenable"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#thenables) (i.e. it has a `.then` function). That's the check which native async/await uses, and means that any kind of promise implementation will work. 

```suggestion
  #isPromise(value) {
    return !!value.then
  }
```

Reply:
good to know! but I actually removed all this code now following our discussion